### PR TITLE
Placeholder for AutoHyperClockCache, more

### PR DESCRIPTION
**Pull request by pdillinger**
Summary:
* The plan is for AutoHyperClockCache to be selected when HyperClockCacheOptions::estimated_entry_charge == 0, and in that case to use a new configuration option min_avg_entry_charge for determining an extreme case maximum size for the hash table. For the placeholder, a hack is in place in HyperClockCacheOptions::MakeSharedCache() to make the unit tests happy despite the new options not really making sense with the current implementation.
* Mostly updating and refactoring tests to test both the current HCC (internal name FixedHyperClockCache) and a placeholder for the new version (internal name AutoHyperClockCache).
* Simplify some existing tests not to depend directly on cache type.
* Type-parameterize the shard-level unit tests, which unfortunately requires more syntax like `this->` in places for disambiguation.
* Added means of choosing auto_hyper_clock_cache to cache_bench, db_bench, and db_stress, including add to crash test.
* Add another templated class BaseHyperClockCache to reduce future copy-paste
* Added ReportProblems support to cache_bench
* Added a DEBUG-level diagnostic to ReportProblems for the variance in load factor throughout the table, which will become more of a concern with linear hashing to be used in the Auto implementation. Example with current Fixed HCC:
```
2023/08/10-13:41:41.602450 6ac36 [DEBUG] [che/clock_cache.cc:1507] Slot occupancy stats: Overall 49% (129008/262144), Min/Max/Window = 39%/60%/500, MaxRun{Pos/Neg} = 18/17
```

In other words, with overall occupancy of 49%, the lowest across any 500 contiguous cells is 39% and highest 60%. Longest run of occupied is 18 and longest run of unoccupied is 17. This seems consistent with random samples from a uniform distribution.

Test Plan: Shouldn't be any meaningful changes yet to production code or to what is tested, but there is temporary redundancy in testing until the new implementation is plugged in.

### CODE REVIEW

File: cache/clock_cache.h [hunk 1]
```diff
@@ -614,10 +621,18 @@ class FixedHyperClockTable : public BaseClockTable {
...
 
+// Placeholder for future automatic table variant
+// For now, just use FixedHyperClockTable.
+class AutoHyperClockTable : public FixedHyperClockTable {
+ public:
+  using FixedHyperClockTable::FixedHyperClockTable;
+};  // class AutoHyperClockTable
+
 // A single shard of sharded cache.
-template <class Table>
```
Comment (jowlyzhang) [inline cache/clock_cache.h:618]: Just for my own learning, is this change and below "using... " a functional change? If so, I wonder why it doesn't work as is.

File: cache/clock_cache.cc [hunk 2]
```diff
@@ -1428,17 +1439,87 @@ void AddShardEvaluation(const FixedHyperClockCache::Shard& shard,
...
+
+  static std::string PercentStr(size_t a, size_t b) {
+    return std::to_string(uint64_t{100} * a / b) + "%";
+  }
+};
+
+template <class Table>
+void BaseHyperClockCache<Table>::ReportProblems(
+    const std::shared_ptr<Logger>& info_log) const {
+  if (info_log->GetInfoLogLevel() <= InfoLogLevel::DEBUG_LEVEL) {
```
Comment (jowlyzhang) [inline cache/clock_cache.cc:1500]: Very minor: I think the marco below `ROCKS_LOG_AT_LEVEL` would also take care of this check? https://github.com/facebook/rocksdb/blob/main/env/env.cc#L897

File: cache/clock_cache.cc [hunk 3]
```diff
@@ -1428,17 +1439,87 @@ void AddShardEvaluation(const FixedHyperClockCache::Shard& shard,
...
 
+bool IsSlotOccupied(const ClockHandle& h) {
+  return (h.meta.load(std::memory_order_relaxed) >> ClockHandle::kStateShift) !=
+         0;
+}
 }  // namespace
 
+// NOTE: GCC might warn about subobject linkage if this is in anon namespace
+template <size_t N = 500>
+class LoadVarianceStats {
```
Comment (jowlyzhang) [inline cache/clock_cache.cc:1450]: This parameter's usage is relatively straightforward, I wonder why it's made a template parameter instead of a simple static class member?

**Review by jowlyzhang — APPROVED** [review]
Looking great! Thank you for the PR @pdillinger. Is the idea that the implementation PR for `AutoHyperClockCache` wouldn't need to update any test or tools, just need to add the implementation and remove the trick in `MakeSharedCache`?

File: cache/clock_cache.h (hunk 1 above)
Comment (pdillinger) [reply cache/clock_cache.h:618]: lru_cache_test uses the Table type, and this change makes it a public member type of the Shard type, so the test can use Shard::Table.

File: cache/clock_cache.cc (hunk 2 above)
Comment (pdillinger) [reply cache/clock_cache.cc:1500]: Computing the data is non-trivial cost, so we don't want to compute it unless it's going to be used. ReportProblems is used in production, but not typically for DEBUG level.

File: cache/clock_cache.cc (hunk 3 above)
Comment (pdillinger) [reply cache/clock_cache.cc:1450]: Just thinking generically I guess

File: cache/clock_cache.cc (hunk 2 above)
Comment (jowlyzhang) [reply cache/clock_cache.cc:1500]: Thanks for the explanation. You are always thinking multiple steps ahead. These ideas are great for me to learn!

**Comment by pdillinger** [comment]
> Looking great! Thank you for the PR @pdillinger. Is the idea that the implementation PR for `AutoHyperClockCache` wouldn't need to update any test or tools, just need to add the implementation and remove the trick in `MakeSharedCache`?

It will need to update tests, but hopefully only in places where the *behavior* of the new cache needs adjustment to the behavior of the test. I don't want those changes drowned out by a sea of changes to the *form* of the tests.
